Adds support for pre-/post- ddev command tasks

tannerjfco opened the pull request:
## The Problem:
OP #183 We would like to provide support for running tasks before or after some ddev commands. These tasks should be defined in the config.yml for a site.

## The Fix:
This PR makes that happen. It adds support for an `extend-commands` key to be defined in the config.yml of a site, that then defines command hooks, with a list of tasks to run. For example:

```
extend-commands:
  post-import-db:
    - exec: "drush cc all"
```

The following command hooks can be defined:
- `pre-start`: Hooks into "ddev start". Execute tasks before the site environment starts.
- `post-start`: Hooks into "ddev start". Execute tasks after the site environment has started
- `pre-import-db`: Hooks into "ddev import-db". Execute tasks before database import
- `post-import-db`: Hooks into "ddev import-db". Execute tasks after database import
- `pre-import-files`: Hooks into "ddev import-files". Execute tasks before files are imported
- `post-import-files`: Hooks into "ddev import-files". Execute tasks after files are imported.

The following tasks can be defined:
- exec: Run a command in the web container
- import-db: Import a database
- import-files: Import files

An example configuration relevant to the apptype of the site is provided in config.yml when "ddev config" is run.

## The Test:
- "ddev config" a new site. Your config.yml should provide an example extend-commands.
- Run a "ddev import-db" to test the example extend-commands. Note: for Wordpress, you'll need to uncomment and provide URLs to replace. The output of import-db should indicate that it is executing post-import commands.
- Review documentation, and test out various configurations.

## Automation Overview:
<!-- Please provide an overview of tests introduced by this PR, or an explanation for why no tests are needed. -->
Tests have been introduced to validate example config is written to config.yml, and to test execution of commands defined in config.

## Related Issue Link(s):
#183 OP

## Release/Deployment notes:
<!-- Does this affect anything else, or are there ramifications for other code? Does anything have to be done on deployment? -->

tannerjfco commented:
This is in a working state, but may have some additional work to follow, depending on some decisions on these criteria items:

> We'll file a separate ticket to run system commands, but not consider it in scope for this one.

I don't think it should be much more effort to add, just need to decide on a key name for the task in yaml. I was thinking maybe "shell", but not tied to anything. The difficulty is limiting potential confusion with "exec". I'll likely add this as part of this PR unless there's a strong objection. If we don't add this, we should probably remove the pre-start hook, since it won't have anything it can successfully run.

> Limit to these 6: start, stop, exec, remove, import-files, import-db

I did not add pre/post hooks or tasks for stop or remove, as I'm not certain of the use case for either. If we want pre/post hooks for them, it's not hard to add.

rickmanelius commented:
@tannerjfco A couple of comments:
* I noticed that in the #183 we separated "post" and "import-db" into two levels of hierarchy and here we merged them into one. Not sure the pros and cons, but noting it as a departure of what was discussed. I figured keeping them split made it more of a generic pattern, but I'm not married to the approach if there is a good reason.
* How about `exec-host`? I agree that this has the potential to get mixed up and/or confused if we go with `exec` versus `shell`, so using `exec` and `exec-host` (less likely to be used) would probably help people keep them all straight. I'm not opposed to adding this is now if it's a short left (~ less than 4 hours or so), but I can't think of a use case quite except perhaps the macOS specific instructions for getting XDebug to work.

tannerjfco commented:
I reduced the hierarchy of the yaml structure primarily to reduce complexity, both for implementation and for users defining tasks. I think this structure is a bit clearer with intent to users as well. I can't think of any cons to this simplified structure over the one originally intended.

rickmanelius commented:
Beginning my tests this morning.

rickmanelius commented:
https://github.com/drud/ddev/pull/348

## Repos Used

* [ddev](https://github.com/drud/ddev/pull/348/commits/604eeca64c4e34a08f335347e6559194adaae47f)
* [Drupal Commerce Kickstart](https://github.com/drud/drupal-kickstart/commit/f980e5053aae1e216b24eb4ca986e40761a8b318)

## Assets Used

* [db.tar.gz](https://github.com/drud/drupal-kickstart/releases/tag/v0.4.0)

## Setup

* Download artifacts from https://circleci.com/gh/drud/ddev/1181#artifacts/containers/0
* ddev_osx.v0.7-21-g604eeca6.zip
* `ddev version`
  * cli:       	v0.7-21-g604eeca6
  * web:       	drud/nginx-php-fpm7-local:v0.7.3
  * db:        	drud/mysql-local-57:v0.5.0
  * dba:       	drud/phpmyadmin:v0.2.0
  * router:    	drud/ddev-router:v0.4.3
  * commit:    	v0.7-21-g604eeca6
  * build info:	Built Mon Jun 26 18:13:47 UTC 2017 circleci@default-1498500658771 drud/golang-build-container:v0.4.2

## Testing

* `rm -rf .ddev` within the drupal-kickstart repo.
* `rm docroot/sites/default/settings.php`
* `ddev config` with defaults
* Confirmed that `more .ddev/config.yml` includes a link to https://ddev.readthedocs.io/en/latest/users/extending-commands/
* Confirmed that a default drupal post-import-db entry is there and commented out.
* Confirmed when using a WordPress site, I can confirm that the example post-import-db example is specific to Wordpress. Thanks!
* `ddev start` to bring up containers
* `ddev import-db --src ~/Downloads/db.tar.gz --extract-path kickstart.sql`
  * Confirmed the import worked as expected.
* Uncommented the last 3 lines in .ddev/config.yml.
  extend-commands:
    post-import-db:
      - exec: drush cc all
* Re-run import-db above.
* I observed the following:
  Executing post-import commands...
  --- Runing exec command: drush cc all ---
  'all' cache was cleared.                                               [success]
  --- Exec command succeeded ---

## Still to Test

* Validation (when a bad command is entered)
* Verifying the actions (I need something easier to confirm than drush cc all, like enabling or disabling a module)

## Recommendations

* Not sure if it's necessary to include the `extend-commands: {}` earlier in the config.yml file. What's the intention of this?

rickmanelius commented:
I'm almost done with my review, but I have a few meetings and wanted to capture current progress. So far so good!

rickmanelius commented:
## Additional Tests

* Extending further
  * Added `- exec: drush dis overlay`
  * Re-ran import
  * Confirmed that I was asked at the prompt to confirm disabling overlay.
    * This gets stuck and doesn't proceed. When I added ` -y` to the end, it worked.
  * Confirmed both commands executed (drush disable overlay and drush cc all)
  * Confirmed I was able to get through the entire process successfully.
* Validation
  * I modified the exec command to `- exec2: drush dis overlay -y`
    * This caused the command to skip.
  * I modified the exec command to `- exec2: drush dis2 overlay -y`
    --- Runing exec command: drush dis2 overlay -y ---
    The drush command 'dis2 overlay' could not be found.  Run `drush         [error]
   cache-clear drush` to clear the commandfile cache if you have installed new extensions.
    Failed to import database for drupal-kickstart: Exec failed: exit status 1

## Recommendations

* Not sure if it's necessary to include the `extend-commands: {}` earlier in the config.yml file. What's the intention of this?
* Commands that involve any sort of interactive prompt cause the process to hang.
* An invalid entry (see `exec2`) might warrant a prompt to the end-user that this is an invalid entry.
* Technically, the import succeeds if an invalid post import-db entry doesn't fire. We should update the error message to be more accurate.
* When the drush command returns an [error], it might be worth highlighting that red to better catch one's attention.
* While we are telling the user that a post-command is executing, we originally discussed alerting a user to the location of where it was firing off. I'm not sure if this is truly valuable or the LOE in providing it, but stating the file + line that triggers it might be of help.

rfay commented:
The inability to accept input using `ddev exec` is not related to this PR, it's a from-the-beginning problem, and should be addressed IMO, but not in this PR.

tannerjfco commented:
> Need to address interactive prompts (see the example with drush dis overlay without the -y flag).

I think this should be addressed in documentation. There is not a good way for us to know what commands should or shouldn't be interactive.

> I suggest we provide a warning to the end-user if they inserted an invalid ddev command (e.g. exec2).

I don't know how we would determine if an arbitrary command will be valid or not. The possibility will exist for users to define invalid commands.

> I suggest we update the error message when a post-hook fails because it currently states the import failed when in fact it was the post import operation that failed.

Error messages bubble up to the parent command, so the parent command will always report as failed. The message can be clarified a bit better indicate it failed because of a pre/post step. An alternative would be to not fail the command if a pre/post step fails and simply output the error, and continue running the other steps / the main command, but it sounded like we wanted to fail entirely if any step fails.

rickmanelius commented:
1. Happy to state "interactive prompts are not supported yet".
2. Wouldn't arbitrary commands be testable against a known list of allowable commands? That is, if I had an array of what ddev functions I'm allowed to run and a specific entry (e.g. `ddev2`) doesn't match, alert the user that it's incorrect? On one hand, it's the user's error. But the silent fail means I'm not likely to even notice it went by.
3. Understood re:bubbling to the parent. From an error tracing perspective, it would be useful to note that there was a fail in the pre/post portions as you suggested. That would at least help for debugging purposes.

rfay commented:
@rickmanelius I think there's a confusion between ddev commands and ddev exec commands? Or I am just missing what's going on. There's decent validation of ddev commands, you can't do a `ddev exec2` - maybe I'm just missing the point.

```
$ ddev exec2
Error: unknown command "exec2" for "ddev"

Did you mean this?
	exec

Run 'ddev --help' for usage.
```

rickmanelius commented:
Hi @rfay. I agree that this validation exists when running a single command. I was referring to if I attempted to add an invalid command in the config.yml here.
```
extend-commands:
  post-import-db:
    - exec2: drush cc all
```
When I do this, it simply ignores the entry and moves on.

rfay commented:
Sorry, yeah yaml validation would be a good thing; it might not be in scope? Definitely would be a good thing everywhere we use yaml.

rickmanelius commented:
@rfay I'm not necessarily looking for full blow validation. But if there is a low LOE manner of flagging issues would be nice. I would imagine a check against a hash of valid commands would be sufficient.

tannerjfco commented:
I have updated this with the following changes:

- The empty `extend-commands: {}` written to config.yaml is gone
- Error messages improved for failed tasks, now specifying what command hook they were in
- Noted in docs for exec/exec-host that interactive commands are not supported
- Added a validation function for extend-commands that checks for valid hooks and valid tasks. I couldn't determine a simplistic or generic way to handle this with the yaml package, as such this validation is specific to the extend-commands key.

rfay commented:
Random worry: `ddev start` can be used in lots of contexts. I use it a lot to just rebuild containers or whatever. I don't think that would be normal for most people, but they could still start an already-started site pretty easily. I guess in general I'm skeptical of attaching things to start at all. However, this is their problem. 
* We do need to give more guidance on the specific situations we want them to use this (import-db for wordpress especially). 
* Some appropriate drupal examples would provide context. For example changing the sitename in import, enabling [environment_indicator](https://www.drupal.org/project/environment_indicator)

rfay commented on pkg/ddevapp/config.go:
```diff
@@ -420,3 +453,60 @@ func prepLocalSiteDirs(base string) error {
 	return nil
 }
+
+// validateCommandYaml validates command hooks and tasks defined in extend-commands for config.yaml
```
Another failure (panic) with ddev import-db (windows):
```
PS C:\Users\randy\workspace\randyfay> ddev import-db --src=randyfay.sql.gz
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x10 pc=0x7ddba8]

goroutine 1 [running]:
github.com/drud/ddev/pkg/plugins/platform.(*LocalApp).GetName(0xc0422c8050, 0x0, 0xcca6c0)
        /go/src/github.com/drud/ddev/pkg/plugins/platform/local.go:151 +0x8
github.com/drud/ddev/cmd/ddev/cmd.glob..func7(0xcff740, 0xc04211cf40, 0x0, 0x1)
        /go/src/github.com/drud/ddev/cmd/ddev/cmd/import-db.go:43 +0x31e
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).execute(0xcff740, 0xc04211cf20, 0x1, 0x1, 0xcff740, 0xc042
11cf20)
        /go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:603 +0x232
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xd00400, 0x0, 0xd00400, 0xc0420b7f50)
        /go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:689 +0x340
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).Execute(0xd00400, 0xc0420b7f68, 0x8dd648)
        /go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:648 +0x32
github.com/drud/ddev/cmd/ddev/cmd.Execute()
        /go/src/github.com/drud/ddev/cmd/ddev/cmd/root.go:98 +0x42
main.main()
        /go/src/github.com/drud/ddev/cmd/ddev/main.go:6 +0x27
PS C:\Users\randy\workspace\randyfay> ddev import-db --src=randyfay.sql.gz
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x10 pc=0x7ddba8]

goroutine 1 [running]:
github.com/drud/ddev/pkg/plugins/platform.(*LocalApp).GetName(0xc0420501b0, 0x0, 0xcca6c0)
        /go/src/github.com/drud/ddev/pkg/plugins/platform/local.go:151 +0x8
github.com/drud/ddev/cmd/ddev/cmd.glob..func7(0xcff740, 0xc04211f030, 0x0, 0x1)
        /go/src/github.com/drud/ddev/cmd/ddev/cmd/import-db.go:43 +0x31e
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).execute(0xcff740, 0xc04211f010, 0x1, 0x1, 0xcff740, 0xc042
11f010)
        /go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:603 +0x232
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xd00400, 0x0, 0xd00400, 0xc0420b7f50)
        /go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:689 +0x340
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).Execute(0xd00400, 0xc0420b7f68, 0x8dd648)
        /go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:648 +0x32
github.com/drud/ddev/cmd/ddev/cmd.Execute()
        /go/src/github.com/drud/ddev/cmd/ddev/cmd/root.go:98 +0x42
main.main()
        /go/src/github.com/drud/ddev/cmd/ddev/main.go:6 +0x27
```

This is with this stanza:
```
extend-commands:
    post-import:
      - exec: "drush vset -y site_name DEVDEV"
      - exec: "drush en -y environment_indicator"
```
(I was explicitly testing the validation on post-import to fail)

rickmanelius replied:
I had a similar issue. When I used the example from before:

```
extend-commands:
  post-import-db:
    - exec2: "drush cc all"
```

Note the issue with `exec2` being an invalid ddev command. I then get the following panic:

```
➜  drupal-kickstart git:(master) ✗ ddev import-db --src ~/Downloads/db.tar.gz --extract-path kickstart.sql
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x13bd168]

goroutine 1 [running]:
github.com/drud/ddev/pkg/plugins/platform.(*LocalApp).GetName(0xc420086030, 0x0, 0x18999c0)
	/go/src/github.com/drud/ddev/pkg/plugins/platform/local.go:151 +0x8
github.com/drud/ddev/cmd/ddev/cmd.glob..func7(0x18cebe0, 0xc42007da80, 0x0, 0x4)
	/go/src/github.com/drud/ddev/cmd/ddev/cmd/import-db.go:43 +0x317
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).execute(0x18cebe0, 0xc42007da40, 0x4, 0x4, 0x18cebe0, 0xc42007da40)
	/go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:603 +0x22b
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x18cf8a0, 0x0, 0x18cf8a0, 0xc42012df50)
	/go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:689 +0x339
github.com/drud/ddev/vendor/github.com/spf13/cobra.(*Command).Execute(0x18cf8a0, 0xc42012df68, 0x14b9b41)
	/go/src/github.com/drud/ddev/vendor/github.com/spf13/cobra/command.go:648 +0x2b
github.com/drud/ddev/cmd/ddev/cmd.Execute()
	/go/src/github.com/drud/ddev/cmd/ddev/cmd/root.go:98 +0x3b
main.main()
	/go/src/github.com/drud/ddev/cmd/ddev/main.go:6 +0x20
```

rickmanelius replied:
I'm on macOS. So this is not specific to Windows.

beeradb commented on docs/users/extending-commands.md:
````diff
@@ -0,0 +1,103 @@
+```
+extend-commands:
+  pre-start:
+    - exec: "git pull origin master"
````
perhaps a composer build or something might be a better example here? I'm worried things that are in documentation may be considered good ideas, and I wouldn't recommend anyone do this, personally.

rickmanelius replied:
Composer seems like a perfect example of when this would be appropriate, particularly if we detect a D8 app type. This was the very thing that @nerdstein had found as something that would trip up a D8 core contributor now that the vendor files are not included in the repo and only available when you download a package.

beeradb commented on docs/users/extending-commands.md:
````diff
@@ -0,0 +1,103 @@
+```
+extend-commands:
+  post-import-db:
+    - import-files:
````
I'm a bit concerned about patterns like this for the following reasons:

1. It could result in looping. Import DB calling import files calling import db ad infinitum.
2. This could potentially wreak havoc with a provider import. The provider import just pulls down archives and then runs each individual (db, file) import one after another. I can certainly imagine instances where this would break things.

Maybe the point of powerful tools is that it allows people to shoot themselves in the foot sometimes, but I do have some reservations in this case.

This critique is restricted to just the import-files and import-db operations. I have no issue w/ the general presence of hooks.

rickmanelius replied:
I would agree with this one. Recursion is a concern and suggesting putting an import within an import could easily lead one to this scenario. I feel there are better examples of post-import functions that would add value (cache clear, user login, etc).

beeradb commented on pkg/ddevapp/config.go:
```diff
@@ -90,6 +106,16 @@ func (c *Config) Write() error {
+	case "wordpress":
+		cfgbytes = append(cfgbytes, []byte(WordPressHooks)...)
+	}
+
```
I feel like doing nice things for applications we know about is OK, but we may want to think about the following:

1. Perhaps find a way to consolidate this app specific logic so it's not so scattered everywhere. Currently it sort of randomly resides all over the place.
2. Make sure we always provide a no-op or "other" that handles applications we don't know about in a sane way.

beeradb replied:
to be clear, I'm also not advocating for changes here. Just spitballing about how we may make it more sustainable going forward.

beeradb commented on pkg/ddevapp/config.go:
```diff
@@ -420,3 +453,60 @@ func prepLocalSiteDirs(base string) error {
 }
+
+// validateCommandYaml validates command hooks and tasks defined in extend-commands for config.yaml
+func validateCommandYaml(source []byte) error {
```
I can imagine situations where we may want this public. Your call, though.

beeradb commented on pkg/plugins/platform/local.go:
```diff
@@ -157,6 +157,14 @@ func (l *LocalApp) ImportDB(imPath string, extPath string) error {
+		if err != nil {
+			return err
+		}
+	}
```
the above code stanza should be a function which accepts a command key and executes the commands. It's repeated ver-batum (besides the key value) a few times.

beeradb commented on pkg/plugins/platform/local.go:
```diff
@@ -416,10 +448,73 @@ func (l *LocalApp) ComposeFiles() []string {
+			if err != nil {
+				return fmt.Errorf("%s database import failed: %v", hookName, err)
+			}
+			util.Success("--- %s database import succeeded ---", hookName)
```
I feel like this sort of messaging should be rolled into the actual import commands. Otherwise, we end up in a situation where output for the same operations varies depending on where/how it is triggered. 

I mostly say this because it bit me some on the pantheon work.

rickmanelius commented on docs/users/extending-commands.md:
````diff
@@ -0,0 +1,103 @@
+        src: "~/Downloads/site-archive.tar.gz"
+        extract-path: "docroot/sites/default/files"
+    - exec: "drush cc all"
+```
````
Agreed. The WordPress import-db item is big enough that it could warrant a prompt during `ddev config`. Alternatively, someone running `import-db` on a WordPress site could be prompted to address this in the config file. There are also situations, notably Pantheon, where we already know what the URL is based on the environment and we can auto generate this.

beeradb replied:
It's worth knowing how we want this to play with providers as well. For instance the pantheon provider could automatically do the find/replace without much grief, and in that case there would be no need for the prompt, since the URL would come straight from pantheon.

beeradb replied:
And for that matter, we wouldn't necessarily need the operation defined in the yaml either. Unless our goal was just to dogfood it.

tannerjfco commented:
Thanks for all the feedback! I believe I've followed up on most of this:

- Improved error messages for config errors
- Updated templates for the example configurations written to the config file, ensuring text is mostly kept to 80 chars wide (doc url exceeds 80)
- Spelling fixes and suggested code cleanup
- Updated the examples in documentation, specifically for pre-start and post-import-db. Pre-start is now a composer example, post-import-db is now a wp search-replace example.

I was not been able to address these:

- `- exec: "drush vset -y site_name \"DEVDEV: randyfay.com\""` The escape slashes are making their way to the exec call, which results in the command not providing the expected result from the drush command. I haven't been able to figure out how to un-escape these quotes so that the command succeeds. strconv.Unquote() seems like it should be the answer, but it would not consider any strings i gave it to be quoted, and would error out. I tried using strings.Replace() as well, but it did not seem to have any affect.
- I was unable to reproduce the panic @rfay noted on Windows on import-db with post-import tasks defined. I tried importing a few different db assets, they all succeeded, and commands ran successfully after. The reported panic looked like it occurred before database import started.
- yaml validation: I haven't found any way to generically validate the yaml before it's read in. I have to expect invalid yaml throws an error on Unmarhsal, the yaml.v2 package doesn't show any way to validate the yaml itself, and I didn't turn up anything useful searching around on the issue. There's also not a good way I see to read in an arbitrary structure and do anything useful with it, so to validate like I'm doing for the extend commands, we have to have some idea of what structure we're reading in to do anything worthwhile. This makes it hard to check for the kind of things we're looking for in the extend commands validation without it being case-specific.

rickmanelius commented:
Ok. I re-reviewed with the following:

## Setup

* Download artifacts from https://circleci.com/gh/drud/ddev/1214#artifacts/containers/0
* ddev_osx.v0.7-31-ga4c10c61.tar.gz
* `ddev version`
  * cli:       	v0.7-31-ga4c10c61

There were 3 items I noted here https://github.com/drud/ddev/pull/348#pullrequestreview-46898753. I note that we solved the interactive prompts by pushing people to the documentation, but unfortunately I'm hitting the same panic issue that @rfay mentioned. Once we resolve that, I'm happy to review again and we should be at an approval. Let me know if this becomes a rabbit hole because I know this feature was something I reqeusted.

tannerjfco commented:
@rickmanelius import-db panic should be resolved now.

rfay commented:
I see these behaviors that may or may not be intentional or adequate:
* `exec: ` with args that should be parsed still doesn't work right, and we may not be able to figure out a workaround. 
* An invalid hook results in "Failed to find an active app" with "Invalid config" add-on. May be good enough. `Failed to find active app to import database to: invalid configuration: invalid command hook post-import defined for extend-commands in config.yaml`
* An invalid action may be ignored  (like with 
```
extend-commands:
    post-import-db:
      - exec: "drush uli"
      - exec3: "drush vset site_name 'DEVDEVDEV randyfay ddev.local'"
```

rfay commented:
There's also still one line too long (91 char) in the .ddev/config.yaml comments. Sorry to be so annoying about that.
```
# command is executed. See https://ddev.readthedocs.io/en/latest/users/extending-commands/
```

rickmanelius commented:
Summarizing:
* [ ] Need to confirm "panic" is gone.
* [ ] Need to clarify/improve "invalid hook" message.
* [ ] Need to line break on long line reference readthedocs
* [ ] Decision/solution on how to deal with escaped arguments with `exec`.

Is that a fair summary / remaining pushlist?

beeradb commented:
I would also add: 

[ ] Needs decision on whether import-db / import-files are appropriate tasks

Personally, I find the usefulness questionable, and it would open up a lot of questions over interoperability with provider plugins

rfay commented:
import-db is the most important by far of all of them @beeradb - import-files unlikely, but there might be call for it.

rfay commented:
- [ ] Formatting of tasks on https://github.com/tannerjfco/ddev/blob/b3fbc762dc70b98b0f9204b30122545a857f6626/docs/users/extending-commands.md is not good enough - needs a quick table probably, and perhaps then the examples (better formatted, not parseable/scannable right now)

tannerjfco commented:
I've removed import-db/files as tasks based on feedback. I also improved the formatting in docs and expanded examples for D7, D8 and WordPress. I believe most feedback has been addressed at this point, just waiting on a successful nightly build run now.

rickmanelius commented:
Okie dokie. Tests passed! Looking for re-reviews from me and @rfay. It might be tricky for me to get mine in tonight, but I'll review (at the latest) Friday morning.

rickmanelius commented:
I'm fine with s/extend-command/hooks/ only because that is how we've been referring to it. That said, "extend commands" would be more understandable to non-devs, but that isn't our target audience for this functionality.

I also won't hold this up if we stick with extend-commands.
